### DISCUSSION TITLE
Custom field with dynamic fields

### DISCUSSION THREAD

**Discussion by dmitrijsmihailovs** [filamentphp/filament#3531] [Help]
I am trying to make a field where you can pass template class, like so
`->template($templateClass)`
and in that templateClass I will have schema method with fields.
I have a problem of getting those fields to show up in form, only solution I found to show them is this code in my field
`$this->afterStateHydrated(function (Component $component, $state) {
            $template = $this->resolveTemplate();
            $this->schema($template->schema());
        });`

But I have a text input field in form and after I type something in there, those template fields disappear.

**Comment by ryangjchandler** [comment]
@dmitrijsmihailovs You'll need to share the code for your custom field as well, otherwise it's hard for us to help you.

**Comment by dmitrijsmihailovs** [comment]
`class CustomField extends Component implements CanEntangleWithSingularRelationships 
{
    use EntanglesStateWithSingularRelationship;

    protected string $view = 'forms::components.fieldset';

    private ?string $template = null;

    final public function __construct(string $label)
    {
        $this->label($label);
    }

    public static function make(string $label): static
    {
        $static = app(static::class, ['label' => $label]);
        $static->configure();

        return $static;
    }

    protected function setUp(): void
    {
        parent::setUp();
        $this->afterStateHydrated(function (Component $component, $state) {
            $template = $this->resolveTemplate();
            $this->schema($template->schema());
        });

        $this->columnSpan('full');
    }

    public function template(?string $template): self
    {
        $this->template = $template;

        return $this;
    }

    private function resolveTemplate()
    {
        $relatedModel = app($this->getRelatedModel());
        if (!$relatedModel instanceof Template) {
            return app($this->template);
        }

        return $relatedModel;
    }
}`

I have based it on Fieldset field and tried to use that "afterStateHydrated" in constructor, make and setUp

**Reply by danharrin** [reply]
How about overriding getChildComponents() to resolve the template? Then you dont need the afterStateHydrated

**Reply by danharrin** [reply]
Or even pass the schema() in inside template()

**Reply by dmitrijsmihailovs** [reply]
Ok, for create looks like this works.
But for edit I get "Typed property Filament\Forms\Components\Component::$container must not be accessed before initialization"
Because I need to use getRelatedModel method to get that template class (that extends Model)

**Reply by danharrin** [reply]
Override `getChildComponents()` then

**Reply by dmitrijsmihailovs** [reply]
Went with another way and got it working, thanks :)

I have another issue, I have a custom list page where I am showing items in tree view, for each item I want to have a button - Add child that opens same modal with filled hidden parentId field.

I have a create action
```
Action::make('create')
                ->mountUsing(function (ComponentContainer $form) {
                    $queryString = $form->getLivewire()->getQueryString();
                    $form->fill([
                        'parentId' => $queryString['parentId'] ?? ''
                    ]);
                })
                ->action(function (array $data): void {
                    $this->redirect(static::$resource::getUrl('create') . '?template=' . $data['template']);
                })
                ->form([
                    Hidden::make('parentId'),
                    Select::make('template')
                        ->label('Template')
                        ->options($this->getTemplates())
                        ->required(),
                ])
```

I can't figure out how to trigger it and pass that parentId parameter

**Reply by danharrin** [reply]
is this action passed into `getActions()` ?

**Reply by dmitrijsmihailovs** [reply]
Yes, it is in getActions()
I found this 
`wire:click="mountAction('create')"`
and it works, but not sure how I would pass that parentId parameter to it.

**Reply by danharrin** [reply]
It you're using `<x-filament::page>` in the page view it should render the action in the top right out of the box

**Reply by dmitrijsmihailovs** [reply]
Yes, that is working. I just want to have that same action trigger, but with additional parameter which I can use in the form, when I click "Add child" button for each item I am rendering in that list pages view

**Reply by danharrin** [reply]
You could maybe store it as a public property on the Livewire component and access it with `$this`?

**Reply by dmitrijsmihailovs** [reply]
```
<a wire:click="mountAction('create')">Add child</a>
```
This triggers that action, problem is I can't figure out how I could pass additional parameter so I can use it in form

**Reply by dmitrijsmihailovs** [reply]
Ok, I found a solution, but not sure if it is a good one :D 

`wire:click="mountAction('create', { 'parentId': {{ $item->getKey() }} })"`

and I overrode mountAction() function

```
public function mountAction(string $name, array $parameters = [])
{
     $this->mountedActionData = array_merge($this->mountedActionData, $parameters);

     parent::mountAction($name);
}
```

Then I added mountUsing to the action

```
->mountUsing(function (ComponentContainer $form) {
      $form->fill($form->getLivewire()->mountedActionData);
})
```

**Comment by Ryuuji159** [comment]
I know is been like 3 years but i'm doing something really similar, how did you solve the problem of `Typed property Filament\Forms\Components\Component::$container must not be accessed before initialization"` when the state gets updated?

**Reply by danharrin** [reply]
Cannot help without a code example
